go/adbc/driver/bigquery/record_reader.go

```diff
@@ -0,0 +1,198 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package bigquery
+
+import (
+	"bytes"
+	"cloud.google.com/go/bigquery"
+	"context"
+	"errors"
+	"github.com/apache/arrow-adbc/go/adbc"
+	"github.com/apache/arrow/go/v16/arrow"
+	"github.com/apache/arrow/go/v16/arrow/array"
+	"github.com/apache/arrow/go/v16/arrow/ipc"
+	"github.com/apache/arrow/go/v16/arrow/memory"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/api/iterator"
+	"sync/atomic"
+)
+
+type reader struct {
+	refCount   int64
+	schema     *arrow.Schema
+	chs        []chan arrow.Record
+	curChIndex int
+	rec        arrow.Record
+	err        error
+
+	cancelFn context.CancelFunc
+}
+
+func deserializeSchema(info []byte, mem memory.Allocator) (*arrow.Schema, error) {
+	rdr, err := ipc.NewReader(bytes.NewReader(info), ipc.WithAllocator(mem))
+	if err != nil {
+		return nil, err
+	}
+	defer rdr.Release()
+	return rdr.Schema(), nil
+}
+
+func checkContext(maybeErr error, ctx context.Context) error {
+	if maybeErr != nil {
+		return maybeErr
+	} else if ctx.Err() == context.Canceled {
+		return adbc.Error{Msg: "Cancelled by request", Code: adbc.StatusCancelled}
+	} else if ctx.Err() == context.DeadlineExceeded {
+		return adbc.Error{Msg: "Deadline exceeded", Code: adbc.StatusTimeout}
+	}
+	return ctx.Err()
+}
+
+// kicks off a goroutine for each endpoint and returns a reader which
+// gathers all of the records as they come in.
+func newRecordReader(ctx context.Context, cl *bigquery.Client, projectID, query string, queryConfig bigquery.QueryConfig, alloc memory.Allocator) (rdr array.RecordReader, err error) {
+	q := cl.Query("")
+	q.QueryConfig = queryConfig
+	q.QueryConfig.Q = query
+	job, err := q.Run(ctx)
+	if err != nil {
+		return nil, err
+	}
+	iter, err := job.Read(ctx)
+	if err != nil {
+		return nil, err
+	}
+	arrowIterator, err := iter.ArrowIterator()
+	if err != nil {
+		return nil, err
+	}
+
+	schema, err := deserializeSchema(arrowIterator.SerializedArrowSchema(), alloc)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Instead of this, why not do:

```go
arrowItReader := bigquery.NewArrowIteratorReader(arrowIterator)
rdr, err := ipc.NewReader(arrowItReader, ipc.WithAllocator(alloc))
```

and then you can just spin through the reader with `for rdr.Next()` ?